### PR TITLE
faster minifier

### PR DESCRIPTION
**Pull request by lemire**
On @jkeiser's new branch (jkeiser/minify) we use 2.15 cycles per input byte to minify...

```
$ git checkout jkeiser/minify
$ make minifiercompetition && ./minifiercompetition jsonexamples/twitter.json 
json_minify                                     :    4.189 cycles per input byte (best)    4.196 cycles per input byte (avg)    0.813 GB/s (error margin: 0.001 GB/s)
simdjson->minify                                :    2.152 cycles per input byte (best)    2.171 cycles per input byte (avg)    1.582 GB/s (error margin: 0.014 GB/s)
minisize = 466906, original size = 631515  (minified down to 73.93 percent of original) 
```

On this new PR (which is on top of jkeiser/minify), we get 0.6 cycles per input byte to minify...

```
$ git checkout dlemire/fastminifier
$ make minifiercompetition && ./minifiercompetition jsonexamples/twitter.json 
json_minify                                     :    4.191 cycles per input byte (best)    4.203 cycles per input byte (avg)    0.813 GB/s (error margin: 0.002 GB/s)
simdjson->minify                                :    0.609 cycles per input byte (best)    0.612 cycles per input byte (avg)    5.576 GB/s (error margin: 0.026 GB/s)
minisize = 466906, original size = 631515  (minified down to 73.93 percent of original) 
```

Of course, results will depend on the dataset and the processor.

This is not yet well tested, it needs to reviewed, but it should be essentially correct.

This would fix https://github.com/lemire/simdjson/issues/384

### CODE REVIEW

**Comment by lemire** [comment]
Credit-wise, this PR is *on top* of jkeiser/minify.

**Comment by lemire** [comment]
(I don't think speeding up the minifier was critical work. The real motivation here is to test out the new design that @jkeiser pushed.)

**Comment by lemire** [comment]
We can now see the stats with all possible implementations...

```
$ ./minifiercompetition jsonexamples/twitter.json 
json_minify                                     :    4.188 cycles per input byte (best)    4.202 cycles per input byte (avg)    0.813 GB/s (error margin: 0.003 GB/s)
simdjson->minify+haswell                        :    0.600 cycles per input byte (best)    0.602 cycles per input byte (avg)    5.662 GB/s (error margin: 0.019 GB/s)
simdjson->minify+westmere                       :    0.787 cycles per input byte (best)    0.788 cycles per input byte (avg)    4.319 GB/s (error margin: 0.006 GB/s)
simdjson->minify+fallback                       :    4.182 cycles per input byte (best)    4.197 cycles per input byte (avg)    0.814 GB/s (error margin: 0.003 GB/s)
```

This is pretty much the kind of result you want.

I also made it so that the minifier tool can take a flag so that you can specify the architecture you want. This might allow someone to sidestep AVX if they so desire.

**Comment by lemire** [comment]
@jkeiser wrote "I bet we could pull 5GB/s". Indeed we can. In fact, we achieve "5.662 GB/s" on the twitter dataset on my test above.

**Comment by jkeiser** [comment]
> Note that this problem becomes trivial (and likely very fast) if you assume AVX-512.

Big time. Taking this algorithm and turning it into something we can reuse will also let us vectorize other things as well, such as copying string bytes.

> I also made it so that the minifier tool can take a flag so that you can specify the architecture you want. This might allow someone to sidestep AVX if they so desire.

I think we should add an environment variable, SIMDJSON_IMPLEMENTATION=<name>, and use that if it's set. That way we don't have to make arguments for all our programs.

**Comment by jkeiser** [comment]
> (I don't think speeding up the minifier was critical work. The real motivation here is to test out the new design that @jkeiser pushed.)

Also agreed. This is sort of a toy problem, and we could likely make it faster too--the current implementation spends some extra instructions finding operators ([]{},:), which is unnecessary for minification. But I don't think we care.

**Comment by jkeiser** [comment]
Oh, thank you for graciously not commenting on my rename of "hamming." I don't know who he is and it's always confused me :)

File: src/arm64/simd.h [hunk 1]
```diff
@@ -142,6 +144,40 @@ namespace simdjson::arm64::simd {
...
       return lookup_table.apply_lookup_16_to(*this);
     }
+
+
+    // Copies to 'output" all bytes corresponding to a 0 in the mask (interpreted as a bitset).
+    // Passing a 0 value for mask would be equivalent to writing out every byte to output.
+    // Only the first 16 - count_ones(mask) bytes of the result are significant but 16 bytes
+    // get written.
+    template<typename L>
+    really_inline void compress(uint16_t mask, L * output) const {
```
Comment (jkeiser) [inline src/arm64/simd.h]: I could imagine having both a `compress(mask)` and `compress(mask, output)` method--the former returning the compressed `simd8<uint8_t>`. (Man, soon enough I'm going to get mad at the number of characters and angle brackets and make a typedef like `simd_u8` / `simd_i8`--I'd really like some analogue of `u8x64` but `u8xmax` looks weird.)

Comment (jkeiser) [reply src/arm64/simd.h]: By "I could imagine," to be clear, I mean "no need to actually do this, this is excellent and totally fine the way it is, it just tickled my brain."

File: src/jsonminifier.cpp [hunk 2]
```diff
@@ -38,6 +36,7 @@ static uint8_t jump_table[256 * 3] = {
     0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1,
 };
 
+// This is a (slow) portable minifier.
```
Comment (jkeiser) [inline src/jsonminifier.cpp]: Assuming we're satisfied with the implementation (I think we are), I'd actually move this into `fallback->minify()` and make the core `json_minify()` method call `active_implementation->minify()`.

**Comment by lemire** [comment]
https://en.wikipedia.org/wiki/Hamming_weight

**Comment by lemire** [comment]
> the current implementation spends some extra instructions finding operators ([]{},:), which is unnecessary for minification. But I don't think we care.

I don't care per se, but are we sure that it does so?

**Comment by lemire** [comment]
This PR:
```
simdjson->minify+haswell                	:    0.474 cycles per input byte (best)    0.483 cycles per input byte (avg)    6.510 GB/s (error margin: 0.116 GB/s)
```

Version of this PR where I destroy the classification code:

```
$ git diff
diff --git a/src/haswell/stage1_find_marks.h b/src/haswell/stage1_find_marks.h
index 6850bb3..696af0c 100644
--- a/src/haswell/stage1_find_marks.h
+++ b/src/haswell/stage1_find_marks.h
@@ -34,10 +34,11 @@ really_inline json_character_block json_character_block::classify(const simd::si
     return _in == simd8<uint8_t>(_mm256_shuffle_epi8(whitespace_table, _in));
   }).to_bitmask();

-  uint64_t op = in.map([&](simd8<uint8_t> _in) {
+  uint64_t op = 0;
+// in.map([&](simd8<uint8_t> _in) {
     // | 32 handles the fact that { } and [ ] are exactly 32 bytes apart
-    return (_in | 32) == simd8<uint8_t>(_mm256_shuffle_epi8(op_table, _in-','));
-  }).to_bitmask();
+  //  return (_in | 32) == simd8<uint8_t>(_mm256_shuffle_epi8(op_table, _in-','));
+ // }).to_bitmask();
   return { whitespace, op };
```

I get...

```
simdjson->minify+haswell                	:    0.475 cycles per input byte (best)    0.582 cycles per input byte (avg)    6.510 GB/s (error margin: 1.210 GB/s)
```


@jkeiser What makes you think that we pay a price of the identification of the structural character? My guess is that with all the inlining, the compiler figures out that it does not need to do this work and does away with it.

Do you have evidence to the contrary?

**Comment by lemire** [comment]
@jkeiser Honestly, I am a bit confused, because I thought that the selling point of your new approach was that I could basically work on top of it and only use what I need without paying for the rest (either in code complexity or performance). That's what I think it does... if you think differently, I'd like to know!

File: src/arm64/simd.h (hunk 1 above)
Comment (lemire) [reply src/arm64/simd.h]: @jkeiser There is a reason it is not coded this way!!! Here is my challenge to you. Go into the haswell version and try to code `compress(mask)`. Don't just stand there: do it. Go! And then, I bet, you'll see the light. :-) I debated whether to document this API choice, and now that it has resurfaced, I will. I will add a comment.

File: src/jsonminifier.cpp (hunk 2 above)
Comment (lemire) [reply src/jsonminifier.cpp]: Ok. I will do that.

**Comment by jkeiser** [comment]
> > the current implementation spends some extra instructions finding operators ([]{},:), which is unnecessary for minification. But I don't think we care.
> 
> I don't care per se, but are we sure that it does so?

On ARM, definitely: we recognize 64 bytes of operators and whitespace in one inseparable mask reduction. Instead, we could find 128 bytes of whitespace just as efficiently. Again, I super don't care and don't think we should fix it: it only applies to minification.

On Intel, I am hoping the optimizer notices it's not used and removes it. (I suspect it does.)

**Comment by jkeiser** [comment]
> @jkeiser Honestly, I am a bit confused, because I thought that the selling point of your new approach was that I could basically work on top of it and only use what I need without paying for the rest (either in code complexity or performance). That's what I think it does... if you think differently, I'd like to know!

Yep. I believe it does, particularly by relying on the optimizer. I really, really don't mean to imply that we need a lower level. Everything *except* minification needs every single bit of code here (and then some!).

I think of it as a token scanner: it will tell you where tokens are and tell you a lot about what characters are inside each one. The biggest point of the approach is that it lets you mix and match how you *use* the tokens. Whether you validate strings, UTF-8, or structure. Even whether/how you actually iterate the tokens, is left to the caller, which is the most important bit for performance: it doesn't attempt to shift cardinality from bytes to tokens. That's where you start getting branchy.

File: src/arm64/simd.h (hunk 1 above)
Comment (jkeiser) [reply src/arm64/simd.h]: Ah, I see it. I keep forgetting that AVX2 is only a partial implementation of 256-bit SIMD.

Comment (lemire) [reply src/arm64/simd.h]: +1

File: src/jsonminifier.cpp (hunk 2 above)
Comment (lemire) [reply src/jsonminifier.cpp]: I just removed json_minify entirely. Well, we still have the implementation, but it is now entirely in fallback, where it belongs.

**Comment by lemire** [comment]
> On ARM, definitely: we recognize 32 bytes of operators and whitespace in one inseparable mask reduction. Instead, we could find 64 bytes of whitespace just as efficiently. Again, I super don't care and don't think we should fix it: it only applies to minification.

I added a note to this effect in the code. I did not remember this difference between x64 and NEON. Still. The performance is good on ARM. We are still minifying at the speed of stage 1. It is hard to beat by a lot.

Because I left a note, if someone cares, there is the documentation and the opportunity to optimize it further. I'll open an issue.

**Comment by lemire** [comment]
Merging.

**Pull request merged**